### DISCUSSION TITLE
Debouncing validation and request cancellation

### DISCUSSION THREAD

**Discussion by johtso** [unpoly/unpoly#607] [Q&A]
So I've been having a play with htmx and form validation on input events, and found that I had to work around issues where user input would be stomped on by the server response if the user continues typing.

I had to cancel any existing request when the user types before triggering a new request.

I see that unpoly automatically cancels requests that target the same element.

What I was wondering is how that interacts with debouncing (https://unpoly.com/watch-options#debouncing-callbacks)

Is there still an edge case where given a long enough debounce duration, a response arrives after a user has typed more content, but before it has triggered a new request? Or do existing requests get cancelled immediately regardless of debounce? If not is there an option to get this behaviour? I struggle to see a situation where you wouldn't want this..

**Comment by triskweline** [comment]
> I had to work around issues where user input would be stomped on by the server response if the user continues typing.

Take a look at [validating while typing](https://unpoly.com/validation#validating-while-typing).

> I see that unpoly automatically cancels requests that target the same element.

The cancelation strategy is used when multiple links or forms target the same element concurrently. No cancelation is used for mutations within a form, e.g. with `[up-validate]` or `[up-watch]`. In this case subsequent requests are queued until an excisting request has finished.

**Reply by johtso** [reply]
Oh interesting.. so if the user is actively typing, the old request will replace the input with a stale value.

So then you can use on-keep to discard the input element in the response and thereby keep any changes the user has made.

But then you still have a problem if the server uses classes on the input element to indicate validation state.

I guess you would then need to use up-on-keep to copy across the attributes.

I'm guessing there isn't a built in way to keep an element, but copy across all attributes from the response except for the value? I see there's a concept of a macro.. that could be neat..
